src/task_manager.cc

```diff
@@ -232,58 +221,59 @@ namespace detail {
 		return need_horizon;
 	}
 
-	task& task_manager::reduce_execution_front(task_ring_buffer::reservation&& reserve, std::unique_ptr<task> new_front) {
+	task& task_manager::reduce_execution_front(std::unique_ptr<task> new_front) {
 		// add dependencies from a copy of the front to this task
 		const auto current_front = m_execution_front;
 		for(task* front_task : current_front) {
 			add_dependency(*new_front, *front_task, dependency_kind::true_dep, dependency_origin::execution_front);
 		}
 		assert(m_execution_front.empty());
-		return register_task_internal(std::move(reserve), std::move(new_front));
+		return register_task_internal(std::move(new_front));
 	}
 
-	void task_manager::set_epoch_for_new_tasks(const task_id epoch) {
+	void task_manager::set_epoch_for_new_tasks(task* const epoch) {
 		// apply the new epoch to buffers_last_writers and last_collective_tasks data structs
 		for(auto& [_, buffer] : m_buffers) {
-			buffer.last_writers.apply_to_values([epoch](const std::optional<task_id> tid) -> std::optional<task_id> {
-				if(!tid) return tid;
-				return {std::max(epoch, *tid)};
+			buffer.last_writers.apply_to_values([epoch](task* const tsk) {
+				if(tsk == nullptr) return tsk;
```

Comment:
```suggestion
				if(tsk == nullptr) return nullptr;
```

Reply:
This probably meant that uninitialized read detection did not work after an epoch was applied?!